Sources/PackageLoading/PackageBuilder.swift

```diff
@@ -548,13 +548,12 @@ public final class PackageBuilder {
                 if fileSystem.isDirectory(path) {
                     return path
                 }
+                if fileSystem.isFile(path) && path.extension == "zip" {
```

Comment:
is the target.type == .binary  in this case too?

Reply:
Right, this check is not strict enough, since the other target types potentially may meet these criteria, fixed that by adding a check for the `.binary` type (also updated the `binaryArtifacts` array description, since it contains not only downloaded artifacts). 
Unfortunately, I don't see how to make it more readable, e.g. by introducing the `target.isLocalArchivedBinary` predicate, but maybe it is already good enough.

I believe this is the right place to look for the matching managed artifact path, just right after the manifest target path is validated. The only thing that I find strange, in the case of binary targets the path validation is performed when the managed artifacts are already in place.

Reply:
I wonder if we flip the order of the if here it would help make this cleaner?

Reply:
That's [what I did](https://github.com/apple/swift-package-manager/pull/3746/commits/b748d4808aad9fd59cc3a2ae874dd55673abfbfa) in the beginning :) but then reverted the change assuming that local binary targets must first validate their path. But now having more context and knowing that [binary target path validation](https://github.com/apple/swift-package-manager/blob/3d48f94174fe484afbb854abb806cac9a51634ed/Sources/Workspace/Workspace.swift#L1805) is performed before the one in `PackageBuilder` I think we can safely flip the order (thanks to @neonichu [for fixing the missing validation crash](https://github.com/apple/swift-package-manager/pull/3737)). 
That actually makes the logic more consistent and clear, emphasizing that all binary target types do not need the additional path validation at the `PackageBuilder` level.
[Done.](https://github.com/apple/swift-package-manager/pull/3746/commits/2fb30110e2d93937b2d8c4c2c52471658b0597be)

Reply:
Just merged the `main` branch in and adapted to the new API, you could try to run the smoke tests again.